### PR TITLE
Move the strongly-typed `Typed` types into the main types

### LINKED ISSUE #67: Why is Emittery.Typed a separate class?
Curious why `Typed` needs to be a separate class? Couldn't the type signature of the main class just be

```typescript
class Emittery<EventDataMap extends Events = Events, EmptyEvents extends EventName = never> {
}
```

to get basically the same result?

### PR DESCRIPTION
**Pull request by airhorns**
I think the extra type is unnecessary, and it's easier to maintain only one of them!

Fixes #67

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -10,7 +10,15 @@ Emittery also accepts an array of strings and symbols as event names.
...
 
-declare class Emittery {
+type EventNameFromDataMap<EventDataMap> = Extract<
+	keyof EventDataMap,
+	EventName
+>;
+
+declare class Emittery<
+	EventDataMap extends Emittery.Events = { [key: string]: unknown },
+	EmptyEvents extends EventName = string
```
Comment (mmkal) [inline index.d.ts:20]: This is a change in the default behaviour. The `Typed` class used `never` as the default. With this change, I think any emittery instance that doesn't explicitly define empty event names will be allowed to emit anything. There should probably be a test for this.

IMO the behaviour should be: allow any string if the non-empty events do. Otherwise default to never. Achievable with some conditional types I think.

File: index.d.ts [hunk 2]
```diff
@@ -368,47 +425,6 @@ declare namespace Emittery {
...
-
-	const emitter = new Emittery.Typed<{value: string}, 'open' | 'close'>();
-
-	emitter.emit('open');
-	emitter.emit('value', 'foo\n');
-	emitter.emit('value', 1); // TS compilation error
-	emitter.emit('end'); // TS compilation error
-	```
-	*/
-	class Typed<EventDataMap extends Events, EmptyEvents extends EventName = never> extends Emittery {
```
Comment (mmkal) [inline index.d.ts:389]: Is it ok to just remove? It's a breaking change which is maybe ok since emittery is v0 but maybe this could be a deprecated passthrough to the main class?

File: index.d.ts (hunk 1 above)
Comment (airhorns) [reply index.d.ts:20]: Yeah, I wasn't exactly sure what the default should be. I went with something permissive to match the way that the node stdlib `EventEmitter` works, and made strict typing opt in by default. My intention with that was to not break importing `Emittery` or using it from JS land with the typescript language service. As you identified it does indeed break `Emittery.Typed`. I think you're right and that I might be able to create a passthrough type that restricts the generics a bit more and matches the old behaviour as well.

File: index.d.ts (hunk 2 above)
Comment (airhorns) [reply index.d.ts:389]: Will try to fix!

File: index.d.ts (hunk 1 above)
Comment (mmkal) [reply index.d.ts:20]: This is probably good enough: 
```suggestion
	EmptyEvents extends EventName = string extends keyof EventDataMap ? string : never
```

When EventDataMap is the default of `Record<string, unknown>`, you get `string`. If it's not the default (i.e. a user-specified map), you get `never`.

File: index.d.ts (hunk 2 above)
Comment (sindresorhus) [reply index.d.ts:389]: This is ok. I plan to make this a breaking change.

File: index.d.ts [hunk 3]
```diff
@@ -303,24 +347,38 @@ declare class Emittery {
...
+	): void;
 
 	/**
 	Clear all event listeners on the instance.
 
 	If `eventName` is given, only the listeners for that event are cleared.
 	*/
-	clearListeners(eventName?: EventNames): void;
+	clearListeners(
+		eventName?: EventNameFromDataMap<EventDataMap> | EmptyEvents
```
Comment (sindresorhus) [inline index.d.ts:373]: `EventNameFromDataMap<EventDataMap> | EmptyEvents` is repeated a lot. Maybe it should be a type alias.

File: readme.md [hunk 4]
```diff
@@ -316,17 +316,22 @@ object.emit('event');
...
+import Emittery = require("emittery");
+
+const emitter = new Emittery<
+	// pass {[eventName]: <eventArgs>} as the first type argument for events that pass data to their listeners
+	{ value: string },
+	// pass a string union type for the event names that don't have data
+	"open" | "close"
+>();
+
+emitter.emit("open");
```
Comment (sindresorhus) [inline readme.md:331]: Use single-quotes

**Comment by airhorns** [comment]
Ok, since we're ok with a breaking change, updated this to:
 - default to having a loosely typed event data map accepting any event name and any event argument
 - allow overriding the event data map with only one type, where each key is an event name and each value is `undefined` if that event has no listener arguments and otherwise is the type of the argument
 - fix linting issues

Thanks for the feedback!

File: index.d.ts (hunk 3 above)
Comment (airhorns) [reply index.d.ts]: Refactored to make all the types simpler!

File: index.d.ts [hunk 5]
```diff
@@ -303,24 +334,31 @@ declare class Emittery {
...
+		[keyof EventData, EventData[keyof EventData]]
+	>;
 
 	/**
 	Remove an `onAny` subscription.
 	*/
-	offAny(listener: (eventName: EventName, eventData?: unknown) => void): void;
+	offAny(
+		listener: <Event extends keyof EventData>(
+			eventName: Event,
```
Comment (sindresorhus) [inline index.d.ts:346]: Shouldn't `Event` be `Name` for consistency?

File: index.d.ts [hunk 6]
```diff
@@ -5,12 +5,15 @@ Symbol event names can be used to avoid name collisions when your classes are ex
...
-type EventNames = EventName | readonly EventName[];
-
-declare class Emittery {
+// Helper type for turning the passed `EventData` type map into a list of string keys that don't require data alongside the event name when emitting. Uses the same trick that `Omit` does internally to filter keys by building a map of keys to keys we want to keep, and then accessing all the keys to return just the list of keys we want to keep.
+type DatalessEventNames<EventData> = {
+	[Key in keyof EventData]: EventData[Key] extends undefined ? Key : never;
+}[keyof EventData];
+
+declare class Emittery<
+	EventData = {[eventName: string]: any},
```
Comment (sindresorhus) [inline index.d.ts:14]:
```suggestion
	EventData = {[eventName: EventName]: any},
```

?

File: readme.md [hunk 7]
```diff
@@ -316,17 +316,31 @@ object.emit('event');
...
+// TS compilation error because `1` isn't assignable to `string`.
+emitter.emit('open', 1);
 
-emitter.emit('open');
-emitter.emit('value', 'foo\n');
-emitter.emit('value', 1); // TS compilation error
-emitter.emit('end'); // TS compilation error
+// TS compilation error because `other` isn't defined in the event data type map.
+emitter.emit('other');
 ```
```
Comment (sindresorhus) [inline readme.md:344]: I think this example should be in the doc comment for `Emittery` too.

File: index.d.ts [hunk 8]
```diff
@@ -303,24 +334,31 @@ declare class Emittery {
...
 	If `eventName` is given, only the listeners for that event are cleared.
 	*/
-	clearListeners(eventName?: EventNames): void;
+	clearListeners(eventName?: keyof EventData): void;
 
 	/**
 	The number of listeners for the `eventName` or all events if not specified.
 	*/
-	listenerCount(eventName?: EventNames): number;
+	listenerCount(eventName?: keyof EventData): number;
```
Comment (sindresorhus) [inline index.d.ts:361]: Maybe make a type alias for `type EventNames = keyof EventData`?

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 6 above)
Comment (airhorns) [reply index.d.ts:14]: Sadly TypeScript doesn't yet support symbols in index signatures, so this doesn't typecheck, otherwise I would. Since it's just the default, you can still override the EventData to support symbol event names. The TS team is gonna add support, it's just really hard apparently. https://github.com/microsoft/TypeScript/issues/1863

File: index.d.ts (hunk 8 above)
Comment (airhorns) [reply index.d.ts:361]: I opted not to do that because to create type aliases from operators on the generics within class definitions like this, you have to use more generic types with defaults. To the best of my knowledge, you can't just do:

```typescript
class Foo<T> {
  type U = keyof T;
  // ...
}
```

you have to do 

```typescript
class Foo<T, U = keyof T> {
  // ...
}
```

For types that are meant to be consumed by end users, which `Emittery` definitely is, I always find this confusing. The type you are going to use has more type parameters than you are supposed to pass, and without looking at the code or knowing that the code is using this trick, it seems unclear. So, I'd say its better to be verbose in the class body and keep the top level user facing type interface clean, but happy to change if you feel strongly.

File: index.d.ts (hunk 5 above)
Comment (airhorns) [reply index.d.ts:346]: Yep, fixed

File: index.d.ts (hunk 8 above)
Comment (airhorns) [reply index.d.ts:361]: Just realized I did this already for the `DatalessEvents` type so I am being kind of inconsistent, oops. I will move this up in the same way.

Comment (airhorns) [reply index.d.ts:361]: AH I remember why this wasn't possible now -- fancy typesystem reasons. Because of the strange no-type-aliases-only-derived-generic-type-defaults thing above, the generic type will usually be the `keyof EventNames` type, but could be instantiated with a different type if the user passed it in. For that reason, TypeScript can't for sure, always access `EventNames[Name]`, because `Name` only defaults to `keyof EventNames`, but could be something else. So, it doesn't type check, so I can't really do it. Sorry for all the confusion!

File: index.d.ts (hunk 6 above)
Comment (sindresorhus) [reply index.d.ts:46]: At least add a TODO that it needs to be updated when that TS issue is fixed.

**Comment by sindresorhus** [comment]
@mmkal Looks good to you now?

File: index.d.ts [hunk 9]
```diff
@@ -5,12 +5,47 @@ Symbol event names can be used to avoid name collisions when your classes are ex
...
+	[Key in keyof EventData]: EventData[Key] extends undefined ? Key : never;
+}[keyof EventData];
+
 
-declare class Emittery {
+/**
+ * Emittery is a strictly typed, fully async EventEmitter implementation. Event listeners can be registered with `on` or `once`, and events can be emitted with `emit`.
+ *
+ * `Emittery` has a generic `EventData` type that can be provided by users to strongly type the list of events and the data passed to the listeners for those events. Pass an interface of {[eventName]: undefined | <eventArg>}, with all the event names as the keys and the values as the type of the argument passed to listeners if there is one, or `undefined` if there isn't.
+ *
```
Comment (sindresorhus) [inline index.d.ts:18]: Remove all the `*` here. Like in the other doc comments.

File: index.test-d.ts [hunk 10]
```diff
@@ -39,3 +39,17 @@ import Emittery = require('.');
 	const ee = new Emittery();
 	expectError(ee.on('anEvent', (data: any, more: any) => undefined));
 }
+
+// Strict typing
```
Comment (mmkal) [inline index.test-d.ts:43]: There should probably be tests for `on`, `off`, `onAny`, `offAny` and really all public methods, which validate that the argument(s) received by listeners have the correct types. It looks like those tests didn't exist before though so conceivably could come in a different PR.

**Review by mmkal — COMMENTED** [review]
Lgtm - made a comment re tests. Couple of other questions I wouldn't consider blocking but worth asking since this is will be a breaking release:

- should the `.Typed` alias be removed from js too to avoid confusion?
- is there a plan to rewrite in typescript and use `tsc` to generate the d.ts file? If doing that might yield other small breaking changes it'd be nice for users if they got bundled.

**Comment by sindresorhus** [comment]
> should the .Typed alias be removed from js too to avoid confusion?

Yes

**Comment by sindresorhus** [comment]
> is there a plan to rewrite in typescript and use tsc to generate the d.ts file? If doing that night yield other small breaking changes it'd be nice for users if they got bundled.

If anyone wanna do the work, sure. But it's not something I plan to work on.

**Comment by sindresorhus** [comment]
@airhorns Can you do https://github.com/sindresorhus/emittery/pull/69#issuecomment-749423374 ?

**Comment by airhorns** [comment]
Yeah, while adding the extended tests I noticed some other issues, still trying to work through those.

**Comment by airhorns** [comment]
Ok, I think this is ready for rereview!

File: index.test-d.ts [hunk 11]
```diff
@@ -1,55 +1,139 @@
...
-import Emittery = require('.');
+import { expectType, expectError } from "tsd";
+import Emittery = require(".");
 
 // emit
 {
 	const ee = new Emittery();
-	ee.emit('anEvent');
-	ee.emit('anEvent', 'some data');
+	ee.emit("anEvent");
```
Comment (sindresorhus) [inline index.test-d.ts:7]: Don't do unrelated changes. Applies to a lot of places in this commit.

Comment (airhorns) [reply index.test-d.ts:7]: I was pretty sure it was actually `xo` doing this which is why I included it, but maybe my editor's format on save option is broken. Should TypeScript be using singles or doubles?

Comment (sindresorhus) [reply index.test-d.ts:7]: Single-quotes. It's not XO that's doing this.

Comment (airhorns) [reply index.test-d.ts:7]: Ah, it is because xo is out of date, which I see you are trying to resolve. Let me know if I can help!

**Comment by sindresorhus** [comment]
Make sure you run `$ npx xo --fix` locally and fix the remaining issues.

**Comment by sindresorhus** [comment]
I did https://github.com/sindresorhus/emittery/pull/69/commits/48ea6ca7857a73317c3cd4e041f5ca563db8cae2 to help you enforce the correct style.

**Comment by airhorns** [comment]
Ok, corrected all my style issues and the latent typescript lint issues, which are unrelated to this PR but resulted from the `xo` version bump.

**Comment by sindresorhus** [comment]
Looks good. Thank you :)
